cmd/kubelet/app/server.go

```diff
@@ -440,6 +449,130 @@ func InitializeTLS(s *options.KubeletServer) (*server.TLSOptions, error) {
 	return tlsOptions, nil
 }
 
+// createCertAndKey tries to:
+// Use the boostrap auth token to get x509 certs if the token is set.
+// If this fails or the token is empty, then self-generates cert and key pair.
+func createCertAndKey(s *options.KubeletServer) error {
+	var err error
+	if s.RequestTLSCert {
+		if err = requestCertFromAPIServer(s); err != nil {
+			// Clean up cert and key.
+			if err := os.Remove(s.TLSCertFile); err != nil {
+				glog.Warningf("Failed to clean up TLS cert file %q: %v", s.TLSCertFile, err)
+			}
+			if err := os.Remove(s.TLSPrivateKeyFile); err != nil {
+				glog.Warningf("Failed to clean up TLS private key file %q: %v", s.TLSPrivateKeyFile, err)
+			}
+			return fmt.Errorf("Cannot get cert from API server: %v", err)
+		}
+
+		glog.V(4).Infof("Using cert got from the API server (%s, %s)", s.TLSCertFile, s.TLSPrivateKeyFile)
+		return nil
+	}
+
+	if err = crypto.GenerateSelfSignedCert(nodeutil.GetHostname(s.HostnameOverride), s.TLSCertFile, s.TLSPrivateKeyFile, nil, nil); err == nil {
+		glog.V(4).Infof("Using self-signed cert (%s, %s)", s.TLSCertFile, s.TLSPrivateKeyFile)
+		return nil
+	}
+	return fmt.Errorf("unable to generate self-signed cert: %v", err)
+}
+
+// requestCertFromAPIServer will:
+// (1) Create a restful client for doing the certificate signing request.
+// (2) Generate key pair and certificate signing request.
+//     The private key is stored to disk.
+// (3) Send request to API server and watch for the issued certificate.
+// (4) Once (3) succeeds, dump the certificate to disk.
+func requestCertFromAPIServer(s *options.KubeletServer) error {
+	// (1).
+	clientConfig, err := CreateAPIServerClientConfig(s)
+	if err != nil {
+		return fmt.Errorf("unable to create API server client config: %v", err)
+	}
+
+	certificatesclient, err := unversionedcertificates.NewForConfig(clientConfig)
+	if err != nil {
+		return fmt.Errorf("unable to create certificates client: %v", err)
+	}
+
+	// (2).
+	hostname := nodeutil.GetHostname(s.HostnameOverride)
+
+	var ips []net.IP
+	nodeIP := net.ParseIP(s.NodeIP)
+	if nodeIP != nil {
+		ips = []net.IP{nodeIP}
+	}
+
+	commonName := fmt.Sprintf("kubelet-%s", hostname)
```

Comment:
@liggitt Also discussed in https://github.com/kubernetes/kubernetes/pull/30094#discussion_r73735338
Adding a prefix basically makes it easier to do the cluster level access control. (But today I am thinking that we could achieve this by making hostname to contain a same prefix)